include/deal.II/lac/schur_complement.h

```diff
@@ -0,0 +1,498 @@
+    const auto S = schur_complement(A_inv,B,C,D);
+    const auto S_prec = PreconditionSelector<...>(D); // D and S operate on same space
+    const auto S_inv = inverse_operator<...>(S,...,prec_S);
+
+    // Solve reduced block system
+    auto rhs = condense_schur_rhs (A_inv,C,f,g); // PackagedOperation that represents the condensed form of g
+    y = S_inv * rhs; // Solve for y
+    x = postprocess_schur_solution (A_inv,B,y,f); // Compute x using resolved solution y
+   @endcode
+ *
+ * In the above example, the preconditioner for \f$ S \f$ was defined as the
+ * preconditioner for \f$ D \f$, which is valid since they operate on the same
+ * space.
+ * However, if \f$ D \f$ and \f$ S \f$ are too dissimilar, then this may lead to
+ * a large number of solver iterations as \f$ \text{prec}(D) \f$ is not a good
+ * approximation for \f$ S^{-1} \f$.
+ *
+ * A better preconditioner would be therefore one that provides a more
+ * representative approximation for \f$ S^{-1} \f$.
+ * This can be achieved by firstly constructing an object that represents an
+ * approximation for \f$ S \f$ wherein expensive operation, namely
+ * \f$ A^{-1} \f$, is approximated.
+ * Thereafter we construct the approximate inverse operator \f$ \tilde{S}^{-1} \f$
+ * which is then used as the preconditioner for computing \f$ S^{-1} \f$.
+ * @code
+    // Construction of approximate inverse of Schur complement
+    const auto A_inv_approx = linear_operator(preconditioner_A);
+    const auto S_approx = schur_complement(A_inv_approx,B,C,D);
+    const auto S_approx_prec = PreconditionSelector<...>(D); // D and S_approx operate on same space
+    const auto S_inv_approx = inverse_operator(S_approx,...,S_approx_prec); // Inner solver: Typically limited to few iterations using IterationNumberControl
+
+    // Construction of exact inverse of Schur complement
+    const auto S = schur_complement(A_inv,B,C,D);
+    const auto S_inv = inverse_operator(S,...,S_inv_approx); // Outer solver
+
+    // Solve reduced block system
+    auto rhs = condense_schur_rhs (lo_A_inv,lo_C,f,g);
+    y = lo_S_inv * rhs; // Solve for y
+    x = postprocess_schur_solution (lo_A_inv,lo_B,y,f);
+   @endcode
+ * Note that due to the construction of @c S_inv_approx and subsequently
+ * @c S_inv, there are a pair of nested iterative solvers which could
+ * collectivly consume a lot of resources.
+ * Therefore care should be taken in the the choices leading to the construction
+ * of the iterative inverse_operators.
+ * One would typically consider the use of a IterationNumberControl
+ * (or a similar mechanism) to limit the number of inner solver iterations.
+ * This controls the accuracy of the approximate inverse operation
+ * \f$ \tilde{S}^{-1} \f$ which acts only as the precondtioner for
+ * \f$ S^{-1} \f$.
+ * Furthermore, the preconditioner to \f$ \tilde{S}^{-1} \f$, which in this example is
+ * \f$ \text{prec}(D) \f$, should ideally be computationally inexpensive.
```

Comment:
@kronbichler Thanks for all of your input, especially regarding the choice of outer solver and solution scheme! I'll get onto making the recommended changes later today :-)

Yes, the idea is that one could construct any preconditioner you like as long as it matches the pattern used by the typical preconditioners. After a quick look at `step-22`, I think that `InverseMatrix` might suffice as it (it looks like `inverse_operator` sends in the preconditioner into the iterative solver "as is"). Otherwise I think that you could wrap this in a `LinearOperator`, something like this:
```c++
SparseILU<double> preconditioner_A;
preconditioner_A.initialize (system_matrix.block(1,1),
                           SparseILU<double>::AdditionalData());
ReductionControl solver_control_M (system_matrix.block(1,1).m(), 1e-10, 1e-6);
SolverCG<> solver_M (solver_control_M);
const auto m_inverse = inverse_operator(system_matrix.block(1,1),solver_M,preconditioner_A);
```
and then use it with the `schur_complement`
```c++
const auto A = linear_operator(system_matrix.block(1,1));
const auto B = linear_operator(system_matrix.block(1,0));
const auto C = linear_operator(system_matrix.block(0,1));
const auto D = null_operator(system_matrix.block(0,0)); // Null operator

const auto A_inv = inverse_operator(A,preconditioner_A);
const auto S = schur_complement(A_inv,B,C,D);

ReductionControl solver_control_S (system_matrix.block(1,1).m(), 1e-10, 1e-6);
SolverCG<> solver_S (solver_control_S);
const auto S_inv = inverse_operator(S,solver_S,m_inverse);

auto x = solution.block(1);
auto y = solution.block(0);
auto f = system_rhs.block(1);
auto g = system_rhs.block(0);
auto rhs = condense_schur_rhs (A_inv,C,f,g);
y = S_inv * rhs;
x = postprocess_schur_solution (A_inv,B,y,f);
```
I think that this is equivalent to what you had done in `step-22`.

Would you like me to construct a reduced case which ensures that this works as expected?